42.md

```diff
@@ -0,0 +1,69 @@
+NIP-42
+======
+
+Authentication of clients to relays
+-----------------------------------
+
+`draft` `optional` `author:Semisol` `author:fiatjaf`
+
+This NIP defines a way for clients to authenticate to relays by signing an ephemeral event.
+
+## Motivation
+
+A relay may want to require clients to authenticate to access restricted resources. For example,
+
+  - A relay may request payment or other forms of whitelisting to publish events -- this can naïvely be achieved by limiting publication
+    to events signed by the whitelisted key, but with this NIP they may choose to accept any events as long as they are published from an
+    authenticated user;
+  - A relay may limit access to `kind: 4` DMs to only the parties involved in the chat exchange, and for that it may require authentication
+    before clients can query for that kind.
+  - A relay may limit subscriptions of any kind to paying users or users whitelisted through any other means, and require authentication.
+
+## Protocol flow
+
+This NIP defines a new message, `AUTH`, which clients can send to relays when they want to authenticate. The message is of the following form:
+
+```
+["AUTH", <signed-event>]
```

Comment:
How do unexpecting relays react to this unknown command? They will probably reply with a notice?

So do I understand correctly that we are using AUTH and not EVENT to avoid this being sent to others who then could use it to auth? Which could only happen if the relay did only occasionally support this nip by maybe having different relay implementations behind some round-robin?

I think we could spare us this new command but tell me please what could go wrong ;)

Reply:
Yes, you should be able to send this command to unexpecting relays and no harm be done. I don't think they will answer with a notice, most likely they will ignore it, but if they do so be it, it is better than having your auth attempt be broadcasted to the world.

I don't see any advantage of using `EVENT` instead of `AUTH`. The relay will already have to have a very specific custom logic for handling auth events anyway, so I think it makes things clearer to have this handled in a separate place already from the get go.

Reply:
I think they shouldn't. Mine doesn't. Or maybe mine does, I don't remember, but I think it doesn't.